eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -0,0 +1,347 @@
+          } else {
+            // Hmm, this is weird
+            // TODO @t-bast : is this ok?
+            Behaviors.unhandled
+          }
+      }
+    }
+
+
+  private def relay(upstream: Upstream.Trampoline, payloadOut: Onion.NodeRelayPayload, packetOut: OnionRoutingPacket): ActorRef = {
+    val paymentCfg = SendPaymentConfig(relayId, relayId, None, paymentHash, payloadOut.amountToForward, payloadOut.outgoingNodeId, upstream, None, storeInDb = false, publishEvent = false, Nil)
+    val routeParams = computeRouteParams(nodeParams, upstream.amountIn, upstream.expiryIn, payloadOut.amountToForward, payloadOut.outgoingCltv)
+    // If invoice features are provided in the onion, the sender is asking us to relay to a non-trampoline recipient.
+    val payFSM = payloadOut.invoiceFeatures match {
+      case Some(features) =>
+        val routingHints = payloadOut.invoiceRoutingInfo.map(_.map(_.toSeq).toSeq).getOrElse(Nil)
+        payloadOut.paymentSecret match {
+          case Some(paymentSecret) if Features(features).hasFeature(Features.BasicMultiPartPayment) =>
+            context.log.debug("relaying trampoline payment to non-trampoline recipient using MPP")
+            val payment = SendMultiPartPayment(adapters.toClassic, paymentSecret, payloadOut.outgoingNodeId, payloadOut.amountToForward, payloadOut.outgoingCltv, nodeParams.maxPaymentAttempts, routingHints, Some(routeParams))
+            val payFSM = fsmFactory.spawnOutgoingPayFSM(context, nodeParams, router, register, paymentCfg, multiPart = true)
+            payFSM ! payment
+            payFSM
+          case _ =>
+            context.log.debug("relaying trampoline payment to non-trampoline recipient without MPP")
+            val finalPayload = Onion.createSinglePartPayload(payloadOut.amountToForward, payloadOut.outgoingCltv, payloadOut.paymentSecret)
+            val payment = SendPayment(adapters.toClassic, payloadOut.outgoingNodeId, finalPayload, nodeParams.maxPaymentAttempts, routingHints, Some(routeParams))
+            val payFSM = fsmFactory.spawnOutgoingPayFSM(context, nodeParams, router, register, paymentCfg, multiPart = false)
+            payFSM ! payment
+            payFSM
+        }
+      case None =>
+        context.log.debug("relaying trampoline payment to next trampoline node")
+        val payFSM = fsmFactory.spawnOutgoingPayFSM(context, nodeParams, router, register, paymentCfg, multiPart = true)
+        val paymentSecret = randomBytes32 // we generate a new secret to protect against probing attacks
+        val payment = SendMultiPartPayment(adapters.toClassic, paymentSecret, payloadOut.outgoingNodeId, payloadOut.amountToForward, payloadOut.outgoingCltv, nodeParams.maxPaymentAttempts, routeParams = Some(routeParams), additionalTlvs = Seq(OnionTlv.TrampolineOnion(packetOut)))
+        payFSM ! payment
+        payFSM
+    }
+    payFSM
+  }
+
+  private def rejectExtraHtlcPartialFunction: PartialFunction[Command, Behavior[Command]] = {
```

Comment:
This is used in only one place, is it really worth putting in a dedicated function (I think it makes it a bit harder to read than if all the `case`s were in the same place)?

Reply:
It's actually used in feature branches.